source/darwin/nw_socket.c

```diff
@@ -0,0 +1,2075 @@
+    } else {
+        // If we are already on event loop or event loop is unavailable, we should already acquire the lock for base
+        // socket access
+        nw_socket->synced_data.base_socket = NULL;
+    }
+
+    aws_ref_count_release(&nw_socket->ref_count);
+    socket->impl = NULL;
+    AWS_ZERO_STRUCT(*socket);
+}
+
+struct read_queue_node {
+    struct aws_allocator *allocator;
+    dispatch_data_t received_data;
+    struct aws_linked_list_node node;
+    size_t current_offset;
+};
+
+static void s_clean_up_read_queue_node(struct read_queue_node *node) {
+    /* releases reference count on dispatch_data_t that was increased during creation of read_queue_node */
+    dispatch_release(node->received_data);
+    aws_mem_release(node->allocator, node);
+}
+
+struct socket_close_complete_args {
+    struct aws_allocator *allocator;
+    aws_socket_on_shutdown_complete_fn *shutdown_complete_fn;
+    void *user_data;
+    struct nw_socket *nw_socket;
+};
+
+static void s_close_complete_callback(struct aws_task *task, void *arg, enum aws_task_status status) {
+    (void)status;
+    struct socket_close_complete_args *task_arg = arg;
+    struct aws_allocator *allocator = task_arg->allocator;
+    if (task_arg->shutdown_complete_fn) {
+        task_arg->shutdown_complete_fn(task_arg->user_data);
+    }
+    aws_ref_count_release(&task_arg->nw_socket->ref_count);
+    aws_mem_release(allocator, task_arg);
+    aws_mem_release(allocator, task);
+}
+
+static void s_socket_impl_destroy(void *sock_ptr) {
+    struct nw_socket *nw_socket = sock_ptr;
+    AWS_LOGF_DEBUG(AWS_LS_IO_SOCKET, "id=%p : start s_socket_impl_destroy", (void *)sock_ptr);
+    /* In case we have leftovers from the read queue, clean them up. */
+    while (!aws_linked_list_empty(&nw_socket->read_queue)) {
+        struct aws_linked_list_node *node = aws_linked_list_pop_front(&nw_socket->read_queue);
+        struct read_queue_node *read_queue_node = AWS_CONTAINER_OF(node, struct read_queue_node, node);
+        s_clean_up_read_queue_node(read_queue_node);
+    }
+
+    /* Network Framework cleanup */
+    if (nw_socket->socket_options_to_params) {
+        nw_release(nw_socket->socket_options_to_params);
+        nw_socket->socket_options_to_params = NULL;
+    }
+
+    aws_socket_on_shutdown_complete_fn *on_cleanup_complete = nw_socket->on_socket_cleanup_complete_fn;
+    void *cleanup_user_data = nw_socket->cleanup_user_data;
+
+    aws_mutex_clean_up(&nw_socket->synced_data.lock);
+    aws_mutex_clean_up(&nw_socket->synced_state.lock);
+    aws_mem_release(nw_socket->allocator, nw_socket);
+
+    nw_socket = NULL;
+
+    if (on_cleanup_complete) {
+        on_cleanup_complete(cleanup_user_data);
+    }
+}
+
+static void s_socket_internal_destroy(void *sock_ptr) {
+    struct nw_socket *nw_socket = sock_ptr;
+    AWS_LOGF_DEBUG(AWS_LS_IO_SOCKET, "id=%p : start s_socket_internal_destroy", (void *)sock_ptr);
+
+    aws_mutex_lock(&nw_socket->synced_data.lock);
```

Comment:
Why is this lock needed?  synced_data protects base_socket which does not seem to be referenced here?